native-authentication/src/main/java/com/okta/nativeauthentication/IdxResponseTransformer.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2022-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.nativeauthentication
+
+import com.okta.authfoundation.client.OidcClientResult
+import com.okta.idx.kotlin.client.InteractionCodeFlow
+import com.okta.idx.kotlin.dto.IdxIdpCapability
+import com.okta.idx.kotlin.dto.IdxPollRemediationCapability
+import com.okta.idx.kotlin.dto.IdxRemediation
+import com.okta.idx.kotlin.dto.IdxResponse
+import com.okta.nativeauthentication.form.Element
+import com.okta.nativeauthentication.form.Form
+import com.okta.nativeauthentication.form.LabelFormBuilder
+import com.okta.nativeauthentication.form.RetryFormBuilder
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.launch
+
+internal class IdxResponseTransformer(
+    private val callback: NativeAuthenticationClient.Callback,
+    private val interactionCodeFlow: InteractionCodeFlow,
+    private val producerScope: ProducerScope<Form>,
+) {
+    suspend fun transformAndEmit(
+        resultProducer: suspend () -> OidcClientResult<IdxResponse>,
+    ) {
+        when (val result = resultProducer()) {
+            is OidcClientResult.Error -> {
+                RetryFormBuilder.emit(producerScope) {
+                    transformAndEmit(resultProducer)
+                }
+            }
+            is OidcClientResult.Success -> {
+                val response = result.result
+                if (response.isLoginSuccessful) {
+                    exchangeInteractionCodeForTokensAndEmit(response)
+                } else {
+                    val builder = Form.Builder()
+                    for (remediation in response.remediations) {
+                        for (field in remediation.form.visibleFields) {
+                            builder.elements.addAll(field.asDynamicAuthFields())
+                        }
+                        builder.elements.addAll(remediation.asAction())
+                    }
+                    producerScope.send(builder.build())
+                }
+            }
+        }
+    }
+
+    private fun IdxRemediation.Form.Field.asDynamicAuthFields(): List<Element.Builder<*>> {
+        return when (true) {
```

Comment:
Isn't this the same as `return when {`?

Reply:
Looks like it! I didn't realize it was optional! Thanks!